apollo-federation/src/query_plan/operation.rs

```diff
@@ -2527,8 +2554,134 @@ impl SelectionSet {
+    ///
+    /// The final selections are optional (for instance, if `path` ends on a leaf field,
+    /// then no followup selections would make sense),
+    /// but when some are provided, uncesssary fragments will be automaticaly removed
+    /// at the junction between the path and those final selections.
+    /// For instance, suppose that we have:
+    ///  - a `path` argument that is `a::b::c`,
+    ///    where the type of the last field `c` is some object type `C`.
+    ///  - a `selections` argument that is `{ ... on C { d } }`.
+    /// Then the resulting built selection set will be: `{ a { b { c { d } } }`,
+    /// and in particular the `... on C` fragment will be eliminated since it is unecesasry
+    /// (since again, `c` is of type `C`).
+    pub(crate) fn add_at_path(
+        &mut self,
+        path: &[Arc<OpPathElement>],
+        selection_set: Option<&Arc<SelectionSet>>,
+    ) -> Result<(), FederationError> {
+        // PORT_NOTE: This method was ported from the JS class `SelectionSetUpdates`. Unlike the
+        // JS code, this mutates the selection set map in-place.
+        match path.split_first() {
+            // If we have a sub-path, recurse.
+            Some((ele, path @ &[_, ..])) => {
+                let mut selection = Arc::make_mut(&mut self.selections)
+                    .entry(ele.key())
+                    .or_insert(|| {
+                        Selection::from_element(
+                            OpPathElement::clone(ele),
+                            // We immediately add a selection afterward to make this selection set
+                            // valid.
+                            Some(SelectionSet::empty(
+                                self.schema.clone(),
+                                self.type_position.clone(),
+                            )),
+                        )
+                    })?;
+                match &mut selection {
+                    SelectionValue::Field(field) => match field.get_selection_set_mut() {
+                        Some(sub_selection) => sub_selection.add_at_path(path, selection_set)?,
+                        None => return Err(FederationError::internal("add_at_path encountered a field without a subselection which should never happen".to_string())),
+                    },
+                    SelectionValue::InlineFragment(fragment) => fragment
+                        .get_selection_set_mut()
+                        .add_at_path(path, selection_set)?,
+                    SelectionValue::FragmentSpread(_fragment) => {
+                        return Err(FederationError::internal("add_at_path encountered a named fragment spread which should never happen".to_string()));
+                    }
+                };
+            }
+            // If we have no sub-path, we can add the selection.
+            Some((ele, &[])) => {
+                if selection_set.is_none() {
```

Comment:
Why don't we delete this if-statement? Since we eagerly updates `self`, there seem to be no differences in the behavior whether or not `selection_set.is_none() && ele.is_terminal()`. Or did I miss something?

Reply:
Good find!